glambie/data/submission_system_interface.py

```diff
@@ -0,0 +1,125 @@
+
+from io import BytesIO
+import json
+import os
+import re
+from typing import List, Optional
+
+from google.cloud import storage
+import pandas as pd
+
+from glambie.const.data_groups import GlambieDataGroup
+from glambie.const.regions import RGIRegion
+
+_storage_client = storage.Client(project='glambie')
+
+_SUBMISSIONS_BUCKET_URI = 'gs://glambie-submissions'
+SUBMISSION_SYSTEM_FLAG = 'glambie_submission_system'
+
+
+def fetch_timeseries_dataframe(user_group: str, region: RGIRegion, data_group: GlambieDataGroup) -> pd.DataFrame:
+    """
+    Download a particular timeseries from the submission system.
+
+    Parameters
+    ----------
+    user_group : str
+        The user group for the submission (called "group_name" within the submission system)
+    region : RGIRegion
+        The RGIRegion for the submission
+    data_group : GlambieDataGroup
+        The GlambieDataGroup for the submission.
+        Note that 'demdiff_and_glaciological' and 'consensus' Data Groups never appear in the submission system.
+
+    Returns
+    -------
+    pd.DataFrame
+        A dataframe containing the loaded submission data. Will have a different format per data_group.
+    """
+    csv_name_in_bucket = "_".join(
+        [region.short_name.lower(),
+         data_group.name.lower().replace("demdiff", "dem_differencing"),
+         user_group.replace(" ", "_").lower()]) + '.csv'
+
+    with BytesIO() as buffer:
+        _storage_client.download_blob_to_file(_SUBMISSIONS_BUCKET_URI + '/' + csv_name_in_bucket, buffer)
+        buffer.seek(0)
+        return pd.read_csv(buffer)
+
+
+def fetch_all_submission_metadata() -> List[dict]:
+    """
+    Download the metadata for all submissions provided to GlaMBIE.
+
+    This fuses the content of both the top-level meta.json and the individual submission metadata
+    files within the submission system, but does not load the Dataset Information File PDF.
+
+    Returns
+    -------
+    List[dict]
+        A list of dictionaries containing the metadata for each GlaMBIE submission.
+        The field values provided for each submission are repeated for each individual Timeseries.
+    """
+    with BytesIO() as buffer:
+        _storage_client.download_blob_to_file(_SUBMISSIONS_BUCKET_URI + '/meta.json', buffer)
+        buffer.seek(0)
+        # note that here, a "dataset" is a single csv file.
+        datasets = json.load(buffer)['datasets']
+
+    # load the other submission metadata.
+    # to save effort, download each submission metadata file only once,
+    # even though we duplicate the information within the Timeseries objects.
+    submission_metadata_buffer = {}
+    for dataset in datasets:
+        if dataset['submission_metadata_filename'] not in submission_metadata_buffer:
+            with BytesIO() as buffer:
+                _storage_client.download_blob_to_file(
+                    _SUBMISSIONS_BUCKET_URI + '/' + dataset['submission_metadata_filename'], buffer)
+                buffer.seek(0)
```

Comment:
The pattern is used 3 times and is almsot all of the bucket interaction. Consider abstracting for json/csv cases

Reply:
right you are, swapped out.